File: web/src/lib/colorCluster.js
import { enhanceImage, lightness, rgbToHex, imageDataToDataURL } from './imageUtils.js'

let colorDb = null

async function loadColorDb() {
  if (colorDb) return colorDb
  const res = await fetch('/colornames.json')
  const payload = await res.json()
  const entries = payload.colors
  colorDb = {
    rgb: entries.map(e => [e.rgb.r, e.rgb.g, e.rgb.b]),
    names: entries.map(e => e.name),
  }
  return colorDb
}

export async function nearestColorName(rgb) {
  const db = await loadColorDb()
  let best = 0
  let bestDist = Infinity
  for (let i = 0; i < db.rgb.length; i++) {
    const dr = db.rgb[i][0] - rgb[0]
    const dg = db.rgb[i][1] - rgb[1]
    const db2 = db.rgb[i][2] - rgb[2]
    const d = dr * dr + dg * dg + db2 * db2
    if (d < bestDist) { bestDist = d; best = i }
  }
  return db.names[best]
}

/**
 * Simple Lloyd's k-means on a Float32Array of shape (N, 3).
 * Returns { centers: number[][], labels: Int32Array }.
 */
function kMeans(pixels, k, maxIter = 20) {
  const n = pixels.length / 3

  // K-means++ seeding
  const usedIdx = new Set()
  const centers = []
  const firstIdx = Math.floor(Math.random() * n)
  usedIdx.add(firstIdx)
  centers.push([pixels[firstIdx * 3], pixels[firstIdx * 3 + 1], pixels[firstIdx * 3 + 2]])

  while (centers.length < k) {
    const dists = new Float32Array(n)
    for (let i = 0; i < n; i++) {
      let minD = Infinity
      for (const c of centers) {
        const dr = pixels[i * 3] - c[0]
        const dg = pixels[i * 3 + 1] - c[1]
        const db = pixels[i * 3 + 2] - c[2]
        const d = dr * dr + dg * dg + db * db
        if (d < minD) minD = d
      }
      dists[i] = minD
    }
    let sum = 0
    for (let i = 0; i < n; i++) sum += dists[i]
    let r = Math.random() * sum
    let chosen = n - 1
    for (let i = 0; i < n; i++) {
      r -= dists[i]
      if (r <= 0) { chosen = i; break }
    }
    centers.push([pixels[chosen * 3], pixels[chosen * 3 + 1], pixels[chosen * 3 + 2]])
  }

  const labels = new Int32Array(n)

  for (let iter = 0; iter < maxIter; iter++) {
    // Assign
    let changed = false
    for (let i = 0; i < n; i++) {
      let bestC = 0, bestD = Infinity
      for (let c = 0; c < k; c++) {
        const dr = pixels[i * 3] - centers[c][0]
        const dg = pixels[i * 3 + 1] - centers[c][1]
        const db = pixels[i * 3 + 2] - centers[c][2]
        const d = dr * dr + dg * dg + db * db
        if (d < bestD) { bestD = d; bestC = c }
      }
      if (labels[i] !== bestC) { labels[i] = bestC; changed = true }
    }
    if (!changed) break

    // Update centers
    const sums = Array.from({ length: k }, () => [0, 0, 0])
    const counts = new Int32Array(k)
    for (let i = 0; i < n; i++) {
      const c = labels[i]
      sums[c][0] += pixels[i * 3]
      sums[c][1] += pixels[i * 3 + 1]
      sums[c][2] += pixels[i * 3 + 2]
      counts[c]++
    }
    for (let c = 0; c < k; c++) {
      if (counts[c] > 0) {
        centers[c] = [sums[c][0] / counts[c], sums[c][1] / counts[c], sums[c][2] / counts[c]]
      }
    }
  }

  return { centers, labels }
}

/**
 * Find dominant colors in an ImageData.
 * Returns array of { rgb, hex, name, count, pct, lightness } sorted by count desc.
 */
export async function findDominantColors(imageData, nColors = 2, saturation = 3.0, contrast = 1.5) {
  const { width, height } = imageData

  // Clone and enhance
  const enhanced = new ImageData(
    new Uint8ClampedArray(imageData.data),
    width,
    height,
  )
  enhanceImage(enhanced, saturation, contrast)

  // Downsample to ~10k pixels for speed
  const total = width * height
  const step = Math.max(1, Math.floor(total / 10000))
  const sampled = []
  for (let i = 0; i < total; i += step) {
    const o = i * 4
    sampled.push(enhanced.data[o], enhanced.data[o + 1], enhanced.data[o + 2])
  }
  const pixels = new Float32Array(sampled)

  const { centers, labels } = kMeans(pixels, nColors)

  // Count pixels per cluster
  const counts = new Int32Array(nColors)
  for (let i = 0; i < labels.length; i++) counts[labels[i]]++

  // Sort by count descending
  const indices = Array.from({ length: nColors }, (_, i) => i)
    .sort((a, b) => counts[b] - counts[a])

  const totalSampled = labels.length
  const results = await Promise.all(
    indices.map(async (ci) => {
      const rgb = centers[ci].map(v => Math.min(255, Math.max(0, Math.round(v))))
      const name = await nearestColorName(rgb)
      return {
        rgb,
        hex: rgbToHex(rgb),
        name,
        count: counts[ci],
        pct: (counts[ci] / totalSampled) * 100,
        lightness: lightness(rgb),
      }
    })
  )

  return results
}

File: color_cluster.py
#!/usr/bin/env python3
"""
Finds dominant colors in an image using k-means clustering, or masks all but
one color cluster to transparent.
Usage:
  python color_cluster.py <image_path> [--colors N] [--saturation F] [--contrast F]
  python color_cluster.py <image_path> --keep <HEX> [--fuzz F]
"""

import sys
import json
import argparse
import urllib.request
from pathlib import Path

import numpy as np
from PIL import Image, ImageEnhance
from sklearn.cluster import KMeans


_COLORNAMES_URL = "https://api.color.pizza/v1/"
_COLORNAMES_CACHE = Path(__file__).parent / "colornames.json"

_db_rgb: np.ndarray | None = None
_db_names: list[str] | None = None


def _load_color_db() -> tuple[np.ndarray, list[str]]:
    global _db_rgb, _db_names
    if _db_rgb is not None:
        return _db_rgb, _db_names

    if not _COLORNAMES_CACHE.exists():
        print("Downloading meodai/color-names database...", file=sys.stderr)
        req = urllib.request.Request(_COLORNAMES_URL, headers={"User-Agent": "color-cluster/1.0"})
        with urllib.request.urlopen(req) as resp:
            _COLORNAMES_CACHE.write_bytes(resp.read())

    with open(_COLORNAMES_CACHE) as f:
        payload = json.load(f)

    entries = payload["colors"]
    names: list[str] = []
    rgb_list: list[list[int]] = []
    for entry in entries:
        c = entry["rgb"]
        rgb_list.append([c["r"], c["g"], c["b"]])
        names.append(entry["name"])

    _db_rgb = np.array(rgb_list, dtype=np.float32)
    _db_names = names
    return _db_rgb, _db_names


def nearest_color_name(rgb: tuple[int, int, int]) -> str:
    db_rgb, db_names = _load_color_db()
    q = np.array(rgb, dtype=np.float32)
    dists = ((db_rgb - q) ** 2).sum(axis=1)
    return db_names[int(np.argmin(dists))]


def find_dominant_colors(
    image_path: str,
    n_colors: int = 2,
    saturation: float = 3.0,
    contrast: float = 1.5,
) -> list[tuple[tuple[int, int, int], int]]:
    img = Image.open(image_path).convert("RGB")
    img = ImageEnhance.Color(img).enhance(saturation)
    img = ImageEnhance.Contrast(img).enhance(contrast)
    pixels = np.array(img).reshape(-1, 3).astype(float)

    kmeans = KMeans(n_clusters=n_colors, random_state=42, n_init=10)
    kmeans.fit(pixels)

    centers = kmeans.cluster_centers_
    labels = kmeans.labels_

    counts = np.bincount(labels)
    sorted_indices = np.argsort(-counts)  # descending by cluster size

    return [
        (tuple(int(round(c)) for c in centers[i]), int(counts[i]))
        for i in sorted_indices
    ]


def rgb_to_hex(rgb: tuple[int, int, int]) -> str:
    return "#{:02X}{:02X}{:02X}".format(*rgb)


def hex_to_rgb(hex_color: str) -> tuple[int, int, int]:
    h = hex_color.lstrip("#")
    return (int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16))


def keep_color(
    image_path: str,
    keep_hex: str,
    fuzz: float,
    saturation: float = 3.0,
    contrast: float = 1.5,
) -> None:
    keep_rgb = hex_to_rgb(keep_hex)
    color_name = nearest_color_name(keep_rgb)
    safe_name = color_name.lower().replace(" ", "-")

    p = Path(image_path)
    out_path = p.parent / f"{p.stem}_{safe_name}_{int(fuzz)}.png"

    img = Image.open(image_path).convert("RGB")
    enhanced = ImageEnhance.Color(img).enhance(saturation)
    enhanced = ImageEnhance.Contrast(enhanced).enhance(contrast)

    pixels = np.array(enhanced.convert("RGBA"))
    rgb = pixels[:, :, :3].astype(float)
    dist = np.sqrt(((rgb - np.array(keep_rgb, dtype=float)) ** 2).sum(axis=2))
    pixels[dist > fuzz, 3] = 0

    Image.fromarray(pixels, "RGBA").save(out_path)
    print(f"Saved: {out_path}")


def lightness(rgb: tuple[int, int, int]) -> float:
    r, g, b = rgb
    luminance = 0.299 * r + 0.587 * g + 0.114 * b
    return (luminance / 255.0) * 100.0


def main():
    parser = argparse.ArgumentParser(
        description="Find dominant colors in an image using k-means clustering."
    )
    parser.add_argument("image_path")
    parser.add_argument("--colors", type=int, default=2, metavar="N")
    parser.add_argument("--saturation", type=float, default=3.0, metavar="F",
                        help="saturation boost factor (default 3.0)")
    parser.add_argument("--contrast", type=float, default=1.5, metavar="F",
                        help="contrast boost factor (default 1.5)")
    parser.add_argument("--keep", metavar="HEX",
                        help="hex color to keep (e.g. #A3B2C1); all other colors become transparent")
    parser.add_argument("--fuzz", type=float, default=30.0, metavar="F",
                        help="max Euclidean RGB distance from --keep color to treat as a match (default 30)")
    args = parser.parse_args()

    image_path = args.image_path

    if args.keep:
        try:
            keep_color(image_path, args.keep, args.fuzz, args.saturation, args.contrast)
        except FileNotFoundError:
            print(f"Error: file not found: {image_path}")
            sys.exit(1)
        except Exception as e:
            print(f"Error: {e}")
            sys.exit(1)
        return

    try:
        colors = find_dominant_colors(
            image_path,
            n_colors=args.colors,
            saturation=args.saturation,
            contrast=args.contrast,
        )
    except FileNotFoundError:
        print(f"Error: file not found: {image_path}")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

    total_pixels = sum(count for _, count in colors)
    print(f"Top {len(colors)} dominant colors in '{image_path}' (sat×{args.saturation}, contrast×{args.contrast}):\n")
    for i, (rgb, count) in enumerate(colors, start=1):
        pct = 100 * count / total_pixels
        name = nearest_color_name(rgb)
        light = lightness(rgb)
        print(f"  {i:>2}. {name:<28}  RGB{rgb}  {rgb_to_hex(rgb)}  lightness={light:.0f}%  ({count:,} px, {pct:.1f}%)")


if __name__ == "__main__":
    main()
